apps/src/applab/applab.js

```diff
@@ -1013,11 +1018,70 @@ Applab.createElement = function (elementType, left, top) {
 
   var divApplab = document.getElementById('divApplab');
   divApplab.appendChild(el);
-  Applab.levelHtml = divApplab.innerHTML;
+  Applab.makeDraggable($(el));
+  Applab.levelHtml = Applab.serializeToLevelHtml();
+};
+
+/**
+ *
+ * @param {jQuery} jq jQuery object containing DOM elements to make draggable.
+ */
+Applab.makeDraggable = function (jq) {
+  var gridSize = 20;
+  jq.draggable({
+    cancel: false,  // allow buttons and inputs to be dragged
+    drag: function(event, ui) {
+      // draggables are not compatible with CSS transform-scale,
+      // so adjust the position in various ways here.
+
+      // dragging
+      var scale = Applab.getVizScaleFactor();
+      var changeLeft = ui.position.left - ui.originalPosition.left;
+      var newLeft  = (ui.originalPosition.left + changeLeft) / scale;
+      var changeTop = ui.position.top - ui.originalPosition.top;
+      var newTop = (ui.originalPosition.top + changeTop) / scale;
+
+      // containment
+      var container = $('#divApplab');
+      var maxLeft = container.width() - ui.helper.outerWidth(true);
+      var maxTop = container.height() - ui.helper.outerHeight(true);
+      newLeft = Math.min(newLeft, maxLeft);
+      newLeft = Math.max(newLeft, 0);
+      newTop = Math.min(newTop, maxTop);
+      newTop = Math.max(newTop, 0);
+
+      // grid
+      newLeft -= newLeft % gridSize;
+      newTop -= newTop % gridSize;
+
+      ui.position.left = newLeft;
+      ui.position.top = newTop;
+    },
+    stop: function(event, ui) {
+      Applab.levelHtml = Applab.serializeToLevelHtml();
+    }
+  });
+};
+
+Applab.getVizScaleFactor = function () {
+  var width = $('body').width();
+  var vizScaleBreakpoints = [1150, 1100, 1050, 1000, 0];
+  if (vizScaleBreakpoints.length !== Applab.vizScaleFactors.length) {
+    throw 'Wrong number of elements in Applab.vizScaleFactors '
+        + Applab.vizScaleFactors;
+  }
+  for (var i = 0; i < vizScaleBreakpoints.length; i++) {
+    if (width > vizScaleBreakpoints[i]) {
+      return Applab.vizScaleFactors[i];
+    }
+  }
+  window.console && console.warn('Unexpected body width: ' + width);
```

Comment:
How unexpected is this? In general, I prefer to see throws for unexpected behavior, since they're more likely to be noticed.